[#113] Add tests for patterns with language conditions

juampynr opened the pull request:
This pull request is for this issue: https://github.com/md-systems/pathauto/issues/113.

TODO:
- [x] Fix logic for bulk-generating aliases for translated content.
- [x] Refactor the logic to add a bundle and a language condition.

Here is further insight about the above two:
## Fix logic for bulk-generating aliases for translated content

There is a weird behavior at the end of the test. When using the `node/article/add` form, each node obtains the correct alias. Then, when we re-generate them, we get odd results. These are the aliases before bulk-deleting:

![selection_002](https://cloud.githubusercontent.com/assets/108130/12080501/66aaea18-b25c-11e5-9c09-8f1db535ea92.png)

Then,  we can see that they are gone after bulk deleting:
![selection_004](https://cloud.githubusercontent.com/assets/108130/12080503/73c6f138-b25c-11e5-9114-95a2d3803937.png)

Finally, this is what we get after bulk updating:
![selection_003](https://cloud.githubusercontent.com/assets/108130/12080506/89efb224-b25c-11e5-9801-2ddc4b6b4e55.png)

I have repeated what the test does manually and could verify that aliases for translations are not created when you bulk delete and bulk update.
## Refactor the logic to add a bundle and a language condition

The logic to add a bundle condition and a language condition is tricky and repetitive. We have it alread at a few places in the module:

```
juampy@Juampy/var/www/drupal8/modules/contrib/pathauto(113-test-multilingual-patterns)$ egrep -rn ContextDefinition . | grep language
./pathauto.install:180:          $new_definition = new ContextDefinition('language', 'Language');
./src/Tests/PathautoUnitTest.php:106:    $new_definition = new ContextDefinition('language', 'Language');
./src/Tests/PathautoLocaleTest.php:128:    $new_definition = new ContextDefinition('language', 'Language');
./src/Tests/PathautoLocaleTest.php:148:    $new_definition = new ContextDefinition('language', 'Language');
./src/Form/PatternEditForm.php:247:        $new_definition = new ContextDefinition('language', 'Language');
```

@Berdir, do you prefer this to be explicit or shall we add "addBundleCondition" and a "addLanguageCondition" methods to the PathautoPattern class?

juampynr committed:
[#113] Add tests for patterns with language conditions

Berdir commented on src/Tests/PathautoLocaleTest.php:
```diff
@@ -27,14 +31,19 @@ class PathautoLocaleTest extends WebTestBase {
+    parent::setUp();
+
+    // Create Article node types.
+    if ($this->profile != 'standard') {
```
You possibly copied this check from somewhere but it doesn't make sense. there are no standard profile tests, it's always testing.

juampynr replied:
Yep, I think I copied it from the node tests. I will remove it.

Berdir commented on src/Tests/PathautoLocaleTest.php:
```diff
@@ -84,5 +93,92 @@ function testLanguageAliases() {
+    \Drupal::service('entity.definition_update_manager')->applyUpdates();
+
+    // Create a pattern for English articles.
+    $pattern = $this->createPattern('node', '/the-articles/[node:title]');
```
I think it would be nice to actually use the UI to create the patterns. Then we have that part of the code tested too.

juampynr replied:
We are already doing that at https://github.com/md-systems/pathauto/blob/8.x-1.x/src/Tests/PathautoUiTest.php#L88 but I see that it does not test language and bundle selection. I will adjust this then.

Berdir commented on src/Tests/PathautoLocaleTest.php:
```diff
@@ -84,5 +93,92 @@ function testLanguageAliases() {
+    $edit = array(
+      'update[canonical_entities:node]' => TRUE,
+    );
+    $this->drupalPostForm('admin/config/search/path/update_bulk', $edit, t('Update'));
```
not surprised that this doesn't work at all :)

We have to update EntityAliasTypeBase::bulkUpdate() to loop over all translations of an entity.

juampynr replied:
Yes, I adjusted that. I will push my latest code now.

juampynr committed:
[#113] Add translation support on bulk alias generation

juampynr commented:
I just pushed a fix for supporting translations in bulk update. My manual tests work but the tests fail at this point:

![selection_001](https://cloud.githubusercontent.com/assets/108130/12090217/2345d4b8-b2ed-11e5-99c4-ef0ffecb9f86.png)

I will continue debugging.

juampynr committed:
[#113] Use the UI to create patterns

juampynr committed:
[#113] Always create the article content type

juampynr commented:
@Berdir I have implemented your suggestsions. I will continue debugging that failing assertion.

juampynr committed:
[#113] Fix failing assertion when testing patterns with languages

juampynr commented:
@Berdir, I have fixed the failing assertion. I changed part the code so it uses the UI instead of trying to do things programmatically. Now the aliases are bulk-updated as expected and therefore tests pass:

![image](https://cloud.githubusercontent.com/assets/108130/12092278/5993248a-b2fc-11e5-8b85-d183b6c4dc61.png)

This is ready for another review.

juampynr committed:
[#113] Remove unneeded statements

juampynr commented:
Hey look, tests passed! Woohoo!

Berdir commented on src/PathautoGenerator.php:
```diff
@@ -257,19 +257,20 @@ protected function getPatternByEntityType($entity_type_id) {
       foreach ($this->getPatternByEntityType($entity->getEntityTypeId()) as $pattern) {
         if ($pattern->applies($entity)) {
-          $this->patterns[$entity->getEntityTypeId()][$entity->id()] = $pattern;
+          $this->patterns[$entity->getEntityTypeId()][$entity->id()][$langcode] = $pattern;
```
Interesting. Wondering if there are other conditions that might result in different patterns for the same entity type. We'll figure out I guess.

Berdir commented on src/Plugin/pathauto/AliasType/EntityAliasTypeBase.php:
```diff
@@ -179,6 +179,16 @@ protected function bulkUpdate(array $ids, array $options = array()) {
       \Drupal::service('pathauto.generator')->updateEntityAlias($entity, 'bulkupdate', $options);
+
+      // Create aliases for the entity translations.
+      if ($entity->isTranslatable()) {
```
I would expect that this can be simplifed. getTranslationLanguages() should always work, even if  it is not translatable, it will simply return only a single one. Remove that check and the FALSE, then you only need a single call to updateEntityAlias(). hasTranslation() should also not be necessay, getTranslationLanguages() will never return a translation that doesn't exist, that's exactly what it does.

juampynr replied:
Good catch!

juampynr committed:
[#113] Simplify bulk update process with translations

juampynr commented on src/Plugin/pathauto/AliasType/EntityAliasTypeBase.php:
```diff
@@ -178,7 +178,11 @@ protected function bulkUpdate(array $ids, array $options = array()) {
     foreach ($entities as $entity) {
-      \Drupal::service('pathauto.generator')->updateEntityAlias($entity, 'bulkupdate', $options);
+      // Update aliases for the entity's default language and its translations.
+      foreach ($entity->getTranslationLanguages() as $langcode => $language) {
```
@Berdir. Done. It's way simpler now :+1:

Berdir commented:
Indeed. Merged, thanks.
